### PR TITLE
feat: type imports

### LINKED ISSUE #151: Add the ability to order types
implemented by #153 

**Is your feature request related to a problem?**
I want my `import type` imports to be ordered at the bottom and not grouped in with my 3rd party module imports.

**Describe the solution you'd like**
Something like `<THIRD_PARTY_MODULES>` but named `<TYPE_IMPORTS>` that we can add to `importOrder`.
```json
{
  "importOrder": ["<TYPE_IMPORTS>", "^[./]"],
  "importOrderSeparation": true,
  "importOrderSortSpecifiers": true
}
```
Before:
```ts
import { getMDXComponent } from 'mdx-bundler/client';
import type { GetStaticPaths, GetStaticProps, NextPage } from 'next';
import Head from 'next/head';
import { useMemo } from 'react';
```
After:
```ts
import { getMDXComponent } from 'mdx-bundler/client';
import Head from 'next/head';
import { useMemo } from 'react';

import type { GetStaticPaths, GetStaticProps, NextPage } from 'next';
```

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by Xenfo**
closes #151 

Config:
```json
{
  "importOrder": ["<THIRD_PARTY_TYPES>", "^[./]", "<TYPE>^[./]"],
  "importOrderSeparation": true
}
```
Before:
```ts
import type { ImportDeclaration } from '@babel/types';
import type { ImportGroups } from './src/types';
import { getImportNodesMatchedGroup } from './src/utils/get-import-nodes-matched-group';
import { abort } from 'process';
```
After:
```ts
import { abort } from 'process';

import type { ImportDeclaration } from '@babel/types';

import { getImportNodesMatchedGroup } from './src/utils/get-import-nodes-matched-group';

import type { ImportGroups } from './src/types';
```

### CODE REVIEW

**Comment by DJTB** [comment]
I'd love to see this as well to specify the location of third-party & local type imports 🙇

**Comment by TomBeckett** [comment]
@ayusharma Is this good to merge?

**Comment by vbylen** [comment]
Let's merge this one!

**Comment by QingyuChai** [comment]
Hi, any update on this PR?

**Comment by Xenfo** [comment]
For everyone that's waiting for the PR to be merged, if you really want to use this feature you guys can build the package locally to use until this PR merged. I know it's impractical but it's the best we can do right now.

```sh
gh repo clone trivago/prettier-plugin-sort-imports
cd prettier-plugin-sort-imports/
gh pr checkout 153
```

Next bump the version in `package.json`. After that run this.

```sh
yarn install
yarn compile
yarn pack
```

Copy the tarball over to your project and change the import for this NPM package in `package.json` to this.

```json
"@trivago/prettier-plugin-sort-imports": "file:./the_path_to_tarball"
```

Then to finish up remove `node_modules` and run the install command of your package manager.

**Comment by quanglam2807** [comment]
Thank you so much.

Patch file for anyone using `patch-package`

`@trivago+prettier-plugin-sort-imports+3.3.0.patch`

```patch
diff --git a/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/constants.js b/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/constants.js
index e662721..3cbbefe 100644
--- a/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/constants.js
+++ b/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/constants.js
@@ -1,6 +1,6 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
-exports.newLineNode = exports.THIRD_PARTY_MODULES_SPECIAL_WORD = exports.newLineCharacters = exports.jsx = exports.typescript = exports.flow = void 0;
+exports.newLineNode = exports.TYPE_SPECIAL_WORD = exports.THIRD_PARTY_TYPES_SPECIAL_WORD = exports.THIRD_PARTY_MODULES_SPECIAL_WORD = exports.newLineCharacters = exports.jsx = exports.typescript = exports.flow = void 0;
 var types_1 = require("@babel/types");
 exports.flow = 'flow';
 exports.typescript = 'typescript';
@@ -11,5 +11,7 @@ exports.newLineCharacters = '\n\n';
  * where the not matched imports should be placed
  */
 exports.THIRD_PARTY_MODULES_SPECIAL_WORD = '<THIRD_PARTY_MODULES>';
+exports.THIRD_PARTY_TYPES_SPECIAL_WORD = '<THIRD_PARTY_TYPES>';
+exports.TYPE_SPECIAL_WORD = '<TYPE>';
 var PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE = 'PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE';
 exports.newLineNode = types_1.expressionStatement(types_1.stringLiteral(PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE));
diff --git a/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/utils/get-import-nodes-matched-group.js b/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/utils/get-import-nodes-matched-group.js
index 3d6bf88..c3a7e5c 100644
--- a/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/utils/get-import-nodes-matched-group.js
+++ b/node_modules/@trivago/prettier-plugin-sort-imports/lib/src/utils/get-import-nodes-matched-group.js
@@ -8,16 +8,42 @@ var constants_1 = require("../constants");
  * @param importOrder
  */
 var getImportNodesMatchedGroup = function (node, importOrder) {
-    var groupWithRegExp = importOrder.map(function (group) { return ({
+    var groupWithRegExp = importOrder
+        .sort(function (a, b) {
+        if (a.startsWith(constants_1.TYPE_SPECIAL_WORD) &&
+            !b.startsWith(constants_1.TYPE_SPECIAL_WORD)) {
+            return -1;
+        }
+        if (!a.startsWith(constants_1.TYPE_SPECIAL_WORD) &&
+            b.startsWith(constants_1.TYPE_SPECIAL_WORD)) {
+            return 1;
+        }
+        return 0;
+    })
+        .map(function (group) { return ({
         group: group,
-        regExp: new RegExp(group),
+        regExp: group.startsWith(constants_1.TYPE_SPECIAL_WORD)
+            ? new RegExp(group.replace(constants_1.TYPE_SPECIAL_WORD, ''))
+            : new RegExp(group),
     }); });
     for (var _i = 0, groupWithRegExp_1 = groupWithRegExp; _i < groupWithRegExp_1.length; _i++) {
         var _a = groupWithRegExp_1[_i], group = _a.group, regExp = _a.regExp;
-        var matched = node.source.value.match(regExp) !== null;
-        if (matched)
-            return group;
+        if (group.startsWith(constants_1.TYPE_SPECIAL_WORD)) {
+            if (node.importKind === 'type') {
+                var matched = node.source.value.match(regExp) !== null;
+                if (matched)
+                    return group;
+            }
+        }
+        else {
+            var matched = node.source.value.match(regExp) !== null;
+            if (matched)
+                return group;
+        }
     }
-    return constants_1.THIRD_PARTY_MODULES_SPECIAL_WORD;
+    return node.importKind === 'type' &&
+        importOrder.find(function (group) { return group === constants_1.THIRD_PARTY_TYPES_SPECIAL_WORD; })
+        ? constants_1.THIRD_PARTY_TYPES_SPECIAL_WORD
+        : constants_1.THIRD_PARTY_MODULES_SPECIAL_WORD;
 };
 exports.getImportNodesMatchedGroup = getImportNodesMatchedGroup;

```

**Comment by popuguytheparrot** [comment]
When it will merged?

**Comment by byara** [comment]
Hey everyone, we'll try to get this out before the year's end.

File: src/constants.ts
Comment (byara) [inline src/constants.ts:17]:
```suggestion
export const TYPE_SPECIAL_WORD = '<TS_TYPES>';
```

File: src/constants.ts
Comment (byara) [inline src/constants.ts:16]:
```suggestion
export const THIRD_PARTY_TYPES_SPECIAL_WORD = '<THIRD_PARTY_TS_TYPES>';
```

File: src/utils/get-import-nodes-matched-group.ts
Comment (byara) [inline src/utils/get-import-nodes-matched-group.ts:39]: Maybe extract this function to a separate file so you can test it?

**Review by byara — CHANGES_REQUESTED** [review]
Thank you for your contribution 🎉 
I left a few comment. Would you be able to take a look?

File: src/utils/get-import-nodes-matched-group.ts
Comment (ayusharma) [inline src/utils/get-import-nodes-matched-group.ts:39]: Just a small improvement from my side:

> The sort() method sorts the elements of an array [in place](https://en.wikipedia.org/wiki/In-place_algorithm) and returns the reference to the same array, now sorted.

```suggestion
    // in place sort
    importOrder
        .sort((a, b) => {
            if (
                a.startsWith(TYPE_SPECIAL_WORD) &&
                !b.startsWith(TYPE_SPECIAL_WORD)
            ) {
                return -1;
            }
            if (
                !a.startsWith(TYPE_SPECIAL_WORD) &&
                b.startsWith(TYPE_SPECIAL_WORD)
            ) {
                return 1;
            }
            return 0;
        })

    const groupWithRegExp =
        importOrder.map((group) => ({
            group,
            regExp: group.startsWith(TYPE_SPECIAL_WORD)
                ? new RegExp(group.replace(TYPE_SPECIAL_WORD, ''))
                : new RegExp(group),
        }));
```

**Review by ayusharma — CHANGES_REQUESTED** [review]
Thank you so much for being patient with the PR. Overall it looks good to me. I suggested a few minor changes. Could you please also add some snapshot tests? ❤️

File: src/utils/get-import-nodes-matched-group.ts
Comment (Xenfo) [reply src/utils/get-import-nodes-matched-group.ts:39]: I've removed the sorting, not sure why I had it in the first place.

Comment (Xenfo) [reply src/utils/get-import-nodes-matched-group.ts:39]: I've removed the sorting, not sure why I had it in the first place.

**Comment by instagibb** [comment]
Any updates on when we will see this PR merged and a new version released?

**Comment by popuguytheparrot** [comment]
any updates?

**Review by ayusharma — APPROVED** [review]

**Comment by ayusharma** [comment]
Thank you for being so patient. We will put this next week when @byara is back.

**Review by byara — APPROVED** [review]

**Comment by ayusharma** [comment]
Hi everyone, Sorry, I had to revert this change due to #211.  This change is still in v4.1.0 and the v4.1.1 revets it. Please give us some more time to release it properly in v5 or the next release.

**Pull request merged**
